Sprint 14

Depechist opened the pull request.

Depechist committed:
Первый экран

- Сверстан интерфейс первого экрана приложения

Depechist committed:
Интерфейс

- Почти закончена работа над интерфейсом главного экрана, и двух модальных экранов AddTrackerVC и NewTrackerVC

Depechist committed:
Интерфейс 2

- Добавлен заголовок для секции на главном экране

Depechist committed:
Интерфейс 3

- Трекеры разнесены на разные секции с разными заголовками

Depechist committed:
Интерфейс 4

- Созданы кнопки выбора типа трекера на экранах NewHabitVC и NewEventVC
- Задана обработка кнопки Отменить на этих экранах

Depechist committed:
Интерфейс 5

- Создан экран с установкой расписания для новой привычки SheduleVC

Depechist committed:
Логика

- Добавлена логика работы кнопки + под трекером, а также учета количества выполненных трекеров в TrackerRecord
- Дополнены структуры Tracker, TrackerCategory, TrackerRecord
- Небольшой рефакторинг

Depechist committed:
Логика 2

- Доработано отображение изменения состояния кнопки +/- под трекером
- Небольшой рефакторинг TrackerModel
- Добавлены моковые ячейки

Depechist committed:
Логика 3

- Моковые данные вынесены в класс DataManager
- В TrackerModel добавлен enum с днями недели
- Модель Трекера дополнена свойством shedule для управления днями недели
- При изменении дня недели в календаре таблица с трекерами обновляется согласно дням недели в структуре каждого из трекеров

Depechist committed:
Логика 4

- Добавлен функционал поиска и отображения ячеек по дате в datePicker, по тексту в строке поиска и кнопке Готово, по тексту в строке поиска в реальном времени

Depechist committed:
Логика 5

- Небольшой рефакторинг
- Исправлен баг с неотображением плейсхолдера на пустом массиве в случае, если при поиске выдается пустой экран коллекции

Depechist committed:
Логика 6

- Добавлен функционал добавления привычки
- Добавлен функционал добавления нерегулярного события
- Теперь отметить привычку или нерегулярное событие как выполненное будущей датой невозможно
- Небольшой рефакторинг

- Задание по 14 спринту выполнено и готово к ревью

amardanov commented on Tracker/DataManager.swift:
```diff
@@ -0,0 +1,31 @@
+
+import Foundation
+
+class DataManager {
```
все классы, от которых нет наследования лучше сделать `final class`

amardanov commented on Tracker/TrackerCell.swift:
```diff
@@ -0,0 +1,222 @@
+    // MARK: - UI ELEMENTS
+    
+    // Верхняя часть ячейки
+    var upperView: UIView = {
```
все UI элементы лучше сделать `private`

amardanov commented on Tracker/TrackerCell.swift:
```diff
@@ -0,0 +1,222 @@
+    // Картинка для кнопки + под трекером
+    var plusImage: UIImage = {
+        let image = UIImage(named: "PlusButton")?.withRenderingMode(.alwaysTemplate) ?? UIImage()
+//        let image = UIImage(named: "PlusButton") ?? UIImage()
```
Эту и 112 строчку убрать

amardanov commented on Tracker/TrackerCell.swift:
```diff
@@ -0,0 +1,222 @@
+    // Отслеживаем нажатие на кнопку под трекером
+    @objc private func plusButtonTapped() {
+        guard let trackerId = trackerId, let indexPath = indexPath else {
+            assertionFailure("No tracker ID")
```
Конструкции по типу `assertionFailure` лучше использовать только в тестах, а отсюда лучше убирать, во избежании неправильной работы приложения

amardanov commented on Tracker/TrackerCreation/NewEventViewController.swift:
```diff
@@ -0,0 +1,178 @@
+    
+    let currentWeekDay = Calendar.current.component(.weekday, from: Date())
+        
+    let navBar = UINavigationBar()
```
Это здесь не нужно будет, если показывать `NewEventViewController` обернуть в `UINavgationController` при показе в файле `AddTrackerViewController`

amardanov commented on Tracker/TrackerCreation/NewEventViewController.swift:
```diff
@@ -0,0 +1,178 @@
+
+class NewEventViewController: UIViewController {
+    
+    var dataManager = DataManager.shared
```
все это `private`
UI элементы тоже

amardanov commented on Tracker/TrackerCreation/NewHabitViewController.swift:
```diff
@@ -0,0 +1,217 @@
+    
+    var schedule = [WeekDay]()
+    
+    let navBar = UINavigationBar()
```
Тоже самое, как и выше

amardanov commented on Tracker/TrackerCreation/NewHabitViewController.swift:
```diff
@@ -0,0 +1,217 @@
+
+class NewHabitViewController: UIViewController {
+    
+    var dataManager = DataManager.shared
```
Также про `private`

amardanov commented on Tracker/TrackerCreation/ScheduleViewController.swift:
```diff
@@ -0,0 +1,148 @@
+
+// MARK: WEEKDAY CELL
+
+class ScheduleTableViewCell: UITableViewCell {
```
в Отдельный файл лучше убрать

amardanov commented on Tracker/TrackerCreation/ScheduleViewController.swift:
```diff
@@ -0,0 +1,148 @@
+
+class ScheduleViewController: UIViewController {
+    
+    var selectedDays = [WeekDay]()
```
Также с `private`

amardanov commented on Tracker/TrackerCreation/ScheduleViewController.swift:
```diff
@@ -0,0 +1,148 @@
+    
+    // MARK: UI ELEMENTS
+    
+    let navBar = UINavigationBar()
```
Тоже самое с navBar

amardanov commented on Tracker/TrackerModel.swift:
```diff
@@ -0,0 +1,63 @@
+    case saturday = 7
+    
+    // Вычисляемое свойство, которое возвращает rawValue текущего дня недели
+    var numberValue: Int {
```
можно не писать это свойство, а напрямую обращаться к перечислению и `.rawValue`

amardanov commented on Tracker/TrackerModel.swift:
```diff
@@ -0,0 +1,63 @@
+
+import UIKit
+
+struct Tracker {
```
Каждая структура и `enum` в отдельный файл и у структур можна убрать `init`. Они уже єсть по-умолчанию

Depechist committed:
Логика 7

Проект доработан после код ревью:

- Исправлена логика работы отметки трекера при изменении даты в datePicker на прошедшую или будущую
- Добавлен Алерт при попытке создания привычки или события без названия
- В TrackersViewController добавлено свойство var currentDate: Date для хранения текущей даты
- Цвет свитчеров на экране Расписания изменен на синий согласно макету
- Небольшой рефакторинг
